src/Code02.DaylightTimeZone.pas

```diff
@@ -36,32 +39,209 @@ interface
-function GetDaylightEnd(const area: string; year: word): TDateTime;
+function IsDaylightSaving(const area: string; year: Word): boolean;
+function GetDaylightStart(const area: string; year: Word): TDateTime;
+function GetDaylightEnd(const area: string; year: Word): TDateTime;
+function MakeValidUrl(const area: string; year: Word): string;
+function GetPageContent(const area: string; year: Word): string;
+procedure SaveDataToRecords(const area: string; year: Word);
+function DataInRecords(const data: string): TJSONObject;
+function GetDaylightFromRecord(const area: string; year: Word): boolean;
+function JsonFromDB: TJSONObject;
+function PageStringToDateTime(const date: string; year: Word): TDateTime;
+function GetDaylightDateFromRecord(const area: string; year: Word; param: string): TDateTime;
 
 implementation
 
 uses
   System.SysUtils,
+  System.IOUtils,
+  System.DateUtils,
+  System.StrUtils,
   Code02.HttpGet;
 
+const
+  cUrlStr = 'https://www.timeanddate.com/time/change/%s?year=%d';
+  cNotDTSdiv = '<div class="alert warning">';
+  cDBfile = 'DBrecords.json';
+  cJsonDTS = 'DTS';
+  cJsonStart = 'start';
+  cJsonEnd = 'end';
+
+function GetDaylightDateFromRecord(const area: string; year: Word; param: string): TDateTime;
+var
+  lJsonObj: TJSONObject;
+  lDateUnix: Integer;
+
+begin
+  Result := 0;
+  lJsonObj := DataInRecords(area + year.ToString);
+  if Assigned(lJsonObj) then
+  begin
+    lJsonObj := JsonFromDB;
+    lDateUnix := lJsonObj.GetValue<Integer>(area + year.ToString + '[0].' + param);
+    Result := UnixToDateTime(lDateUnix);
+  end;
+end;
+
+function PageStringToDateTime(const date: string; year: word): TDateTime;
+var
+  lDateParts: TArray<string>;
+  lMonth, lDay, lHour, lMinute: word;
+  lFs: TFormatSettings;
+
+begin
+  lFs := TFormatSettings.Create('en-US');
+  lDateParts := SplitString(date, ' ');
+  lMonth := IndexText(lDateParts[2], lFs.LongMonthNames) + 1;
+  lDay := lDateParts[1].ToInteger;
+  lHour := StrToInt(Copy(lDateParts[3], 1, 2));
+  lMinute := StrToInt(Copy(lDateParts[3], 4, 2));
+  Result := EncodeDateTime(year, lMonth, lDay, lHour, lMinute, 0, 0);
+end;
+
+function JsonFromDB: TJSONObject;
+begin
+  Result := TJSONObject.Create;
+  if FileExists(cDBfile) then
+    Result := TJSONObject(TJSONObject.ParseJSONValue(TFile.ReadAllText(cDBfile)));
+end;
+
+function GetDaylightFromRecord(const area: string; year: word): boolean;
+var
+  lJsonObj: TJSONObject;
+
+begin
+  Result := False;
+  lJsonObj := DataInRecords(area + year.ToString);
+  if Assigned(lJsonObj) then
+  begin
+    lJsonObj := JsonFromDB;
+    Result := lJsonObj.GetValue<boolean>(area + year.ToString + '[0].' + cJsonDTS);
+  end;
+end;
+
+function DataInRecords(const data: string): TJSONObject;
+var
+  lJsonObj: TJSONObject;
+
+begin
+  lJsonObj := JsonFromDB;
+  Result := TJSONObject(lJsonObj.FindValue(data));
+  lJsonObj.Free;
+end;
+
+procedure SaveDataToRecords(const area: string; year: word);
```

Comment:
Ta metoda jest za dużo - wymaga zdekoponowania

Reply:
Nie do końca miałem pomysł na to dekomponowanie. To jest ta część kodu, gdzie wiem że powinno to być poprawione ale nie do końca pomysł jak.